d2-mode.el

```diff
@@ -122,14 +122,161 @@
     (org-babel-eval cmd "")
     nil))
 
-(defun d2--locate-declaration (str)
+(defun d2--locate-declaration (str tag)
   "Locate a certain declaration and return the line difference and indentation.
-STR is the declaration."
+STR is the declaration.
+TAG is a symbol, which is prepended to the declaration data."
   (let ((l (line-number-at-pos)))
     (save-excursion
       (if (re-search-backward str (point-min) t)
-          (cons (- l (line-number-at-pos)) (current-indentation))
-        (cons -1 -1)))))
+          (cons tag (cons (- l (line-number-at-pos)) (current-indentation)))
+        (cons tag (cons -1 -1))))))
+
+(defun d2--combine-tokens-by-level (tokens)
+  "Combine TOKENS that are at the same level.
+Collect the tags from both into a list.
+Use the newer token's line and column."
+  (reverse
+   (cl-reduce (lambda (acc el)
+                (if (and (not (null acc))
+                         (equal (d2--decl-line (car acc))
+                                (d2--decl-line el)))
+                    ;; when tokens are on the same line, combine them
+                    (progn
+                      (setf (car acc)
+                            ;; we don't care about duplicate tags
+                            (cons (cons (d2--decl-tag el) (d2--decl-tag (car acc)))
+                                  (cons (d2--decl-line el)
+                                        (d2--decl-column el))))
+                      acc)
+                  ;; otherwise add a new token; set its tag to be a list
+                  (cons (cons (list (d2--decl-tag el))
+                              (cons (d2--decl-line el)
+                                    (d2--decl-column el)))
+                        acc)))
+              tokens
+              :initial-value ())))
+
+(defun d2--sort-tokens-by-line (tokens)
+  "Sort TOKENS by line number."
+  (sort tokens
```

Comment:
@MoisMoshev I ran this locally, and also ran into this problem. 3 arguments, but only accepts two. Can you clean this up?

Reply:
Hmm I'm confused by this, sort takes just one positional, and key arguments:
`(sort SEQ &key KEY LESSP REVERSE IN-PLACE)`
So it looks ok...? I don't think this changed recently. I'm running emacs 30.0.50

Reply:
Nevermind, just checked in Emacs 29 and indeed the signature changed. It should be 
(sort SEQ PREDICATE)
brb

Reply:
Fixed

Reply:
Love it. Thank you.